### PR TITLE
#6090 Document accessibility considerations

### PR DESCRIPTION
**Pull request by thibaudcolas**
See #6090. This is a new documentation page, meant as a reference of all things to consider for the accessibility of Wagtail sites – essentially raising awareness of common issues we see in countless audits of real-world sites.

### CODE REVIEW

**Comment by thibaudcolas** [comment]
@Scotchester this should now be ready for review – I hope it all makes sense!

File: docs/advanced_topics/accessibility_considerations.rst [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+Ideally, always add an optional “alt text” field wherever an image is used, alongside the image field:
+
+- For normal fields, add an alt text field to your image’s panel.
+- For StreamField, add an extra field to your image block.
+- For rich text – Wagtail already makes it possible to customize alt text for rich text images. Unfortunately it’s not currently possible to elect not to provide alt text for decorative images (see `#6494 <https://github.com/wagtail/wagtail/issues/6494>`_).
+
+When defining the alt text fields, make sure they are optional so editors can choose to not write any alt text for decorative images. Take the time to provide ``help_text`` with appropriate guidance.
+For example, linking to `established resources on alt text <https://axesslab.com/alt-texts/>`_.
+
+.. note:: Why not have an alt text field on the Image model?
```
Comment (chiaberry) [inline docs/advanced_topics/accessibility_considerations.rst:36]: Is this note presenting the case why one shouldnt add an optional "alt text" field or is it a note to explain why Wagtail's image model doesnt have one out of the box?

File: docs/advanced_topics/accessibility_considerations.rst [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+-------------------
+
+The default behavior for Wagtail images is to use the ``title`` field as the alt text (`#4945 <https://github.com/wagtail/wagtail/issues/4945>`_).
+This is inappropriate, as it’s not communicated in the CMS interface, and the image upload form uses the image’s filename as the title by default.
+
+Ideally, always add an optional “alt text” field wherever an image is used, alongside the image field:
+
+- For normal fields, add an alt text field to your image’s panel.
+- For StreamField, add an extra field to your image block.
+- For rich text – Wagtail already makes it possible to customize alt text for rich text images. Unfortunately it’s not currently possible to elect not to provide alt text for decorative images (see `#6494 <https://github.com/wagtail/wagtail/issues/6494>`_).
```
Comment (chiaberry) [inline docs/advanced_topics/accessibility_considerations.rst:31]: "it’s not currently possible to opt out of providing alt text" removes the double negative from the sentence.

or 

"it’s not currently possible to set alt text as an optional field in these situations"

File: docs/advanced_topics/accessibility_considerations.rst (hunk 1 above)
Comment (thibaudcolas) [reply docs/advanced_topics/accessibility_considerations.rst:36]: More the former, but good point, I should reword this so it’s as clear as possible.

File: docs/advanced_topics/accessibility_considerations.rst [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+help_text and HelpPanel
+-----------------------
+
+Occasional Wagtail users may not be aware of your site’s content guidelines, or best practices of writing for the web. Use fields’ ``help_text`` and ``HelpPanel`` (see :doc:`/reference/pages/panels`).
+
+Readability
+-----------
+
+Readability is fundamental to content being accessible. One of the ways to improve content is to have a clear target for reading level / reading age, which can be assessed with `wagtail-readinglevel <https://github.com/vixdigital/wagtail-readinglevel>`_ as a score displayed in rich text fields.
```
[the remark this answers is not in the document]
Comment (allcaps) [reply docs/advanced_topics/accessibility_considerations.rst:147]: Should we use `texts` with an `s`?

**Comment by allcaps** [comment]
Should this guide also mention the obvious: Write valid HTML?

File: docs/advanced_topics/accessibility_considerations.rst [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+By default, Wagtail stores its bold formatting as a ``b`` tag, and italic as ``i`` (`#4665 <https://github.com/wagtail/wagtail/issues/4665>`_). While those tags don’t necessarily always have correct semantics (``strong`` and ``em`` are more ubiquitous), there isn’t much consequence for screen reader users, as by default screen readers do not announce content differently based on emphasis.
+
+If this is a concern to you, you can change which tags are used when saving content with :ref:`rich text format converters <rich_text_format_converters>`. In the future, :ref:`rich text rewrite handlers <rich_text_rewrite_handlers>` should also support this being done without altering the storage format (`#4223 <https://github.com/wagtail/wagtail/issues/4223>`_).
+
+TableBlock
+----------
+
+The :doc:`/reference/contrib/table_block` default implementation makes it too easy for end users to miss they need either row or column headers (`#5989 <https://github.com/wagtail/wagtail/issues/5989>`_). Make sure to always have either row headers or column headers set.
+Also treat its Caption field as mandatory, so screen reader users navigating the site’s tables know where they are.
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply docs/advanced_topics/accessibility_considerations.rst:80]: I’ll rewrite this slightly because I think I could’ve used more plain language to start with.

File: docs/advanced_topics/accessibility_considerations.rst [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+- Follow `official W3C guidance on accessible forms development <https://www.w3.org/WAI/tutorials/forms/>`_.
+
+----
+
+.. _authoring_accessible_content:
+
+Authoring accessible content
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here are things you can do to help content authors take accessibility into consideration.
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply docs/advanced_topics/accessibility_considerations.rst:131]: Thank you, I like it, will change to `create` rather than `compose`.

```suggestion
Here are things you can do to help authors create accessible content.
```

File: docs/advanced_topics/accessibility_considerations.rst [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+In both rich text and custom StreamField blocks, it’s sometimes easy for editors to create a heading block but not add any content to it. If this is a problem for your site,
+
+- Add validation rules to those fields, making sure the page can’t be saved with the empty headings.
+- Consider adding similar validation rules for rich text fields.
+
+Additionally, you can hide empty heading blocks with CSS:
+
+.. code-block:: css
+
+    h1:empty, h2:empty, h3:empty, h4:empty, h5:empty, h6:empty {
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply docs/advanced_topics/accessibility_considerations.rst:109]: Good question – this has been discussed before in #4431, and I don’t know what to recommend. I don’t think there are accessibility concerns with empty paragraphs. I need to check for list items.

File: docs/advanced_topics/accessibility_considerations.rst [hunk 7]
```diff
@@ -0,0 +1,163 @@
+Accessibility considerations
+============================
+
+Accessibility for CMS-driven websites is a matter of :ref:`modeling content appropriately <content_modeling>`, :ref:`creating accessible templates <accessibility_in_templates>`, and :ref:`authoring content <authoring_accessible_content>` with readability and accessibility guidelines in mind.
+Wagtail generally puts developers in control of content modeling and front-end markup, but there are a few areas to be aware of nonetheless, and ways to help authors be aware of readability best practices.
```
Comment (allcaps) [inline docs/advanced_topics/accessibility_considerations.rst:5]: Maybe reword/extend this introduction and make clear this page is about what Wagtail can do to make content accessible. Other topics (contrast, font-size, valid markup) are important but not discussed.

File: docs/advanced_topics/accessibility_considerations.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (allcaps) [reply docs/advanced_topics/accessibility_considerations.rst:109]: Is this a candidate to be handled by the CMS? If empty headings and empty paragraphs are bad. Why are they not filtered out? Of raise a validation error on save? Why would we promote hacks?

[the remark this answers is not in the document]
Comment (thibaudcolas) [reply docs/advanced_topics/accessibility_considerations.rst:109]: Empty paragraphs aren’t bad for accessibility so let’s leave this to #4431. Empty headings are bad, so I’m going to add documentation on how to make sure StreamField heading blocks are required. For rich text, I have opened #6526.

I think promoting the CSS trick is appropriate in this case, because I assume it’s an issue that affects most / all Wagtail sites right now, and there simply is no other ready-to-go solution to it.

File: docs/advanced_topics/accessibility_considerations.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply docs/advanced_topics/accessibility_considerations.rst:150]: I will rephrase a bit but keep "readability" because this isn’t exactly the same as plain language.

**Comment by thibaudcolas** [comment]
Thank you all for the feedback!
